lib/spack/spack/mirror.py

```diff
@@ -495,18 +493,18 @@ def error(self):
         self.errors.add(self.current_spec)
 
 
-def add_single_spec(spec, mirror_root, mirror_stats):
+def add_single_spec(spec, mirror, mirror_stats):
     tty.msg("Adding package {pkg} to mirror".format(
         pkg=spec.format("{name}{@version}")
     ))
     num_retries = 3
     while num_retries > 0:
         try:
             with spec.package.stage as pkg_stage:
-                pkg_stage.cache_mirror(mirror_stats)
+                pkg_stage.cache_mirror(mirror, mirror_stats)
                 for patch in spec.package.all_patches():
-                    if patch.cache():
-                        patch.cache().cache_mirror(mirror_stats)
+                    if patch.stage:
+                        patch.stage.cache_mirror(mirror, mirror_stats)
```

Comment:
Here we might be mirroring patches with a known hash without mirroring the main package. This together with the message that gets printed:
```console
$ spack mirror create --skip-unstable-versions -d /tmp/mirror quantum-espresso@develop
[ ... ]
==> Successfully created mirror in file:///tmp/mirror
  Archive stats:
    0    already present
    1    added
    0    failed to fetch.
```
might confuse the user ("Successfully created mirror"). I'm really undecided if this should be solved here or as part of something like #12796 since the PR works correctly as only resources with a known hash are downloaded.

Reply:
`spack.mirror.create` calls `MirrorCache.next_spec` once for each call to `_add_single_spec`, so the `MirrorCache` object could be updated to understand and report more details when there is a partial success